client/js/self-check.js

```diff
@@ -0,0 +1,240 @@
+/* CONSTANTS */
+const protocol = window.location.protocol + "//";
+const hostname = window.location.hostname;
+const port = window.location.port;
+const baseURL = port ? (protocol + hostname + ":" + port) : (protocol + hostname);
+
+// This is some kind of clever black magic that makes "document" work in functions
+// The "|| {}" lets it fail on .focus() or .value without blowing up functions
+const $ = query => document.querySelector(query) || {}
+
+const hide = (el) => {
+	el.classList.add("hide")
+}
+const show = (el) => {
+	el.classList.remove("hide")
+}
+
+let COVID_SAFE = false
+const configure_for_covid_safe = () => {
+	$("#idinstruction").textContent = "Please scan your ID:"
+	$("#submit").remove()
+	$("#logoutinstruction").textContent = "Scan your ID again to logout"
+	$("#iteminstruction").textContent = "Please scan your next item:"
+}
+
+async function initiate() {
+	getModalBox();
+
+	$("#barcode").onkeyup = (e) => {
+		if (e.key === "Enter") { // e.code = e.key = "Enter" == e.which = 13
+			loan()
+		}
+	}
+
+	$("#userid").onkeyup = (e) => {
+		if (e.key === "Enter") { // e.code = e.key = "Enter" == e.which = 13
+			login()
+		}
+	}
+
+	const response = await fetch("/isCovidSafe").then(r => r.json())
+	if ("covidSafe" in response) {
+		COVID_SAFE = response.covidSafe
+		if (COVID_SAFE) {
+			configure_for_covid_safe()
+		}
+	} else {
+		console.error("Failed to query `/isCovidSafe` endpoint")
+	}
+}
+
+var modal;
+var span;
+var user;
+var sessiontimer;
+var defaultTimeout = 60
+
+function getModalBox() {
+	hide($("#myModal"))
+
+	// When the user clicks on <span> (x), close the modal
+	$("#modalclosex").onclick = () => {
+		hide($("#myModal"))
+	}
+}
+
+function returnToBarcode() {
+	$("#barcode").disabled = false;
+	hide($("#myModal"))
+
+	$("#barcode").value = ""
+	$("#barcode").focus()
+}
+
+
+/* LOGIN */
+
+function login() {
+	let loginid = $("#userid").value
+	if ((loginid != null) && (loginid != "")) {
+
+		$("#userid").disabled = true
+		hide($("#loginerror"))
+
+		$("#modalheader").textContent = "Loading data, please wait..."
+		show($("#myModal"))
+		hide($("#modalclosex"))
+
+		// console.log(baseURL+"/users/"+loginid)
+		fetch(baseURL + "/users/" + $("#userid").value)
+			.then(r => r.json())
+			.then((data) => {
+				user = data
+				console.log(data)
+				// prepare scan box
+				$("#scanboxtitle").textContent = "Welcome " + data.pref_first_name + " " + data.pref_last_name

```

Comment:
Not everyone has a preferred name set. We need to prefer the preferred names, but if they don't exist we should use the official first and last name.

Reply:
Should be fixed now. I'll test and make a new comment if there are any issues.